hy/compiler.py

```diff
@@ -2029,7 +2030,9 @@ def compile_function_def(self, expression):
                     name=name,
                     args=args,
                     body=body.stmts,
-                    decorator_list=[])
+                    decorator_list=[],
+                    **({} if docstring is None else
+                        {'docstring': str_type(docstring)}))
```

Comment:
Just doing `docstring=str_type(docstring)` should work. The ast module's API is basically all `self, /, *args, **kwargs`, and drops unknown `kwargs`. Might make this stuff a little simpler to maintain.

Reply:
Actually, I tested it on your branch. All tests passed on py27 and py36. py37 fails for a few tests (running a patched astor, its not that) - but the macro tests start passing. Haven't dug into it yet.

It also seems like Python 3 always sets docstring to None as a default, so its probably best to match that behaviour.

```python
>>> print(astor.dump_tree(ast.parse('''
... def test():
...     pass
... ''')))
Module(
    body=[
        FunctionDef(name='test',
            args=arguments(args=[], vararg=None, kwonlyargs=[], kw_defaults=[], kwarg=None, defaults=[]),
            body=[Pass],
            decorator_list=[],
            returns=None,
            docstring=None)],
    docstring=None)
```

Reply:
I'm getting no test failures locally with a patched astor.

Reply:
Just to be clear, I mean testing it with a modification on top of your patch to always set `docstring`, not sure if that was clear.

Reply:
Try it without your modification, just with my PR and a fixed astor, and see if you still get test failures.

Reply:
I have `test_bin_hy_byte_compile` failing, with or without my changes, on py37 only. Everything else is working. Weird...

Okay, so something that needs to get investigated but doesn't need to block this PR.

I can't reproduce the failures from Travis either.

Reply:
Aaaah, I fixed the bytecode serializer for PEP 552, but I forgot to fix the loader.

I patched it: https://github.com/vodik/hy/commit/34929ec7ba3252fdc13fa5c59b7dbd01fd0af1de.patch

Feel free to borrow it if you want, or, write it as you see fix.

Reply:
I used `docstring=(None if docstring is None else str_type(docstring))` so that empty docstrings are still recognized.

I thought all the tests were passing because I forgot to run `test_bin`. I added your patch and everything should be good now (except for astor).

Reply:
Nice. If you notice from the above snippet ^ python also sets `docstring=None` on the Module, so you can get rid of that spat trick everywhere.

Reply:
Right, oops. Done.